packages/flutter/lib/src/material/icon_button.dart

```diff
@@ -869,21 +896,25 @@ class _SelectableIconButton extends StatefulWidget {
   final Widget child;
   final VoidCallback? onLongPress;
   final ValueChanged<bool>? onHover;
+  final MaterialStatesController? statesController;
 
   @override
   State<_SelectableIconButton> createState() => _SelectableIconButtonState();
 }
 
 class _SelectableIconButtonState extends State<_SelectableIconButton> {
-  late final MaterialStatesController statesController;
+  late final MaterialStatesController _internalStatesController;
+
+  MaterialStatesController get statesController =>
+      widget.statesController ?? _internalStatesController;
 
   @override
   void initState() {
     super.initState();
     if (widget.isSelected == null) {
-      statesController = MaterialStatesController();
+      _internalStatesController = MaterialStatesController();
```

Comment:
I think we only need to create an `_internalStatesController` when there is not a user provided one.